apis/common/v1/condition.go

```diff
@@ -107,6 +107,21 @@ type ConditionedStatus struct {
 	// +listMapKey=type
 	// +optional
 	Conditions []Condition `json:"conditions,omitempty"`
+
+	// ObservedGeneration is the most recent resource metadata.generation
+	// observed by the reconciler
+	// +optional
+	ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
+
+	// ObservedLabels are the most recent resource metadata.labels
+	// observed by the reconciler
+	// +optional
+	ObserverLabels map[string]string `json:"observedLabels,omitempty"`
+
+	// ObservedLabels are the most recent resource metadata.annotations
+	// observed by the reconciler
+	// +optional
+	ObserverAnnotations map[string]string `json:"observedAnnotations,omitempty"`
```

Comment:
```suggestion
	// ObservedLabels are the most recent resource metadata.labels
	// observed by the reconciler
	// +optional
	ObservedLabels map[string]string `json:"observedLabels,omitempty"`

	// ObservedLabels are the most recent resource metadata.annotations
	// observed by the reconciler
	// +optional
	ObservedAnnotations map[string]string `json:"observedAnnotations,omitempty"`
```

Can you please create a new struct (something like `ObservedStatus`) for these fields, instead of adding them to `ConditionedStatus`. The idea of `ConditionedStatus` is that you can embed it in a status struct to support status conditions. The new struct should be in a different file (not condition.go).

Is there precedent for other tools having observedLabels and observedAnnotations as well as observedGeneration? I'm supportive of observedGeneration but not convinced that it's worth two additional fields for labels and annotations.

Reply:
@negz great that you spot the typo!

> The new struct should be in a different file (not condition.go).

Done, the struct lives now in `apis/common/v1/observation.go`

> Is there precedent for other tools having observedLabels and observedAnnotations as well as observedGeneration? I'm supportive of observedGeneration but not convinced that it's worth two additional fields for labels and annotations.

Value in `metadata.generation` get increased only when resource `spec` changes. Given that our controllers react also on label/annotation changes and often use/propagate them during the reconciliation, writing down just the observed generation might not be enough, i.e. an observer would not be able to detect if/when a label/annotation change got processed.

Reply:
> Do we know of concrete situations (ideally issues) that recording observed labels and annotations would have helped with? If not, I would suggest we omit them until there's evidence they're needed.

The PR adds an ability to record the observed labels/annotations and the both fields are optional. IMHO a controller should set them if they are used anyhow to influence the reconcilation logic.

In a Crossplane claim, its annotations and labels are copied down to its counterpart composite as a part of reconciliation.

If we do not record the observed claim annotations/lables, then an observer is not able to detect if a reconciliation has happened:
* A claim got created and became ready
* The claim annotation got added/removed/updated
* Given that `metadata.generation` did not change, `status.observedGeneration` does not change either after the reconcile

I have found an old issue about asking to increase `metadata.generation` on annotation/label changes:
https://github.com/kubernetes/kubernetes/issues/67428 and very interesting followup discussion.

@sttts what are your thought on this? Should we stop propagating annotations/labels from claim to composite? Should it be acceptable if claim reconciliation on pure annotation/label changes is not detectable by an observer?

Reply:
> If we do not record the observed claim annotations/lables, then an observer is not able to detect if a reconciliation has happened.

I get that, but I'm challenging whether this has ever been a real problem for someone. Should we wait to see if this is hurting people before we fix it?

Reply:
> Should we stop propagating annotations/labels from claim to composite? 

I don't think we can - this would be a breaking behaviour change.

Reply:
> The PR adds an ability to record the observed labels/annotations and the both fields are optional. IMHO a controller should set them if they are used anyhow to influence the reconcilation logic.
> In a Crossplane claim, its annotations and labels are copied down to its counterpart composite as a part of reconciliation.
> If we do not record the observed claim annotations/lables, then an observer is not able to detect if a reconciliation has happened

I tend to agree with these statements.

That having said, IMO passing down labels/annotations from claim to xrd and similar is wrong, especially for annotations. But as they are passed down today, it sounds logical that the controller records them as part of the observed state.